thrust/thrust/complex.h

```diff
@@ -150,374 +45,168 @@ struct complex_storage;
-  complex<T>& operator/=(const U& z);
-
-
-
-  /* --- Getter functions ---
-   * The volatile ones are there to help for example
-   * with certain reductions optimizations
-   */
-
-  /*! Returns the real part of this \p complex.
-   */
-  __host__ __device__
-  T real() const volatile { return data.x; }
-
-  /*! Returns the imaginary part of this \p complex.
-   */
-  __host__ __device__
-  T imag() const volatile { return data.y; }
-
-  /*! Returns the real part of this \p complex.
-   */
-  __host__ __device__
-  T real() const { return data.x; }
-
-  /*! Returns the imaginary part of this \p complex.
-   */
-  __host__ __device__
-  T imag() const { return data.y; }
-
-
-
-  /* --- Setter functions ---
-   * The volatile ones are there to help for example
-   * with certain reductions optimizations
-   */
-
-  /*! Sets the real part of this \p complex.
-   *
-   *  \param re The new real part of this \p complex.
-   */
-  __host__ __device__
-  void real(T re) volatile { data.x = re; }
-
-  /*! Sets the imaginary part of this \p complex.
-   *
-   *  \param im The new imaginary part of this \p complex.e
-   */
-  __host__ __device__
-  void imag(T im) volatile { data.y = im; }
-
-  /*! Sets the real part of this \p complex.
-   *
-   *  \param re The new real part of this \p complex.
-   */
-  __host__ __device__
-  void real(T re) { data.x = re; }
-
-  /*! Sets the imaginary part of this \p complex.
-   *
-   *  \param im The new imaginary part of this \p complex.
-   */
-  __host__ __device__
-  void imag(T im) { data.y = im; }
-
-
+  __host__ operator std::complex<T>() const { return std::complex<T>(this->real(), this->imag()); }
+};
 
-  /* --- Casting functions --- */
+// The following code enables operators between `complex` when the underlying type of the `complex`
+// differ. Using enable_if ensures that these operators are only called, in the case
+// the types differ. In all other cases the operators from the base class ::cuda::std::complex
+// should be called.
+// This is done to avoid the construction of temporary objects which must be done in some of
+// the operators. Also it is avoided to call the implementation of the base class operator
+// in case the operator is trivially implementable, for the same reason.
 
-  /*! Casts this \p complex to a <tt>std::complex</tt> of the same type.
-   */
-  __host__
-  operator std::complex<T>() const { return std::complex<T>(real(), imag()); }
-
-private:
-  typename detail::complex_storage<T, sizeof(T) * 2>::type data;
-};
+/* --- Equality Operators --- */
 
+/*! Returns true if two \p complex numbers with different underlying type
+ * are equal and false otherwise.
+ *
+ *  \param x The first \p complex.
+ *  \param y The second \p complex.
+ *
+ *  \tparam \c T0 is convertible to \c T1.
+ */
+template <typename T0, typename T1>
+__host__ __device__ typename detail::disable_if<detail::is_same<T0, T1>::value, bool>::type
+operator==(const complex<T0> &x, const complex<T1> &y);
```

Comment:
Could we avoid this by making `thrust::complex` an alias to `cuda::std::complex`? Or is there custom logic that we need to implement directly in the `thrust::complex` type that `cuda::std::complex` doesn't provide?

Reply:
Unfortunately, the custom logic is needed for backwards compatibility.
This is also the reason the unit test for `thrust::complex` were extended in a separate PR (https://github.com/NVIDIA/cccl/pull/413) to test that this change is BC and also does not introduce additional functionality. Sorry, should have mentioned that in the PR description.